File: rust/cmd/doubaospeech/src/commands/media.rs
//! Media processing CLI commands

use clap::{Args, Subcommand};
use serde::{Deserialize, Serialize};

use giztoy_doubaospeech::{Language, SubtitleFormat, SubtitleRequest};

use super::{create_client, get_context, load_request, output_result, print_success, print_verbose};
use crate::Cli;

/// Media processing service
#[derive(Args)]
pub struct MediaCommand {
    #[command(subcommand)]
    command: MediaSubcommand,
}

#[derive(Subcommand)]
enum MediaSubcommand {
    /// Extract subtitles from media
    Subtitle {
        /// Media file URL
        #[arg(short = 'u', long)]
        url: Option<String>,
        /// Output format (srt, vtt, txt)
        #[arg(short = 'F', long)]
        format: Option<String>,
        /// Language
        #[arg(short = 'l', long)]
        language: Option<String>,
    },
    /// Query media task status
    Status {
        /// Task ID to query
        task_id: String,
    },
}

/// Subtitle request from YAML/JSON file
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
struct SubtitleFileRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    media_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    format: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    language: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    callback_url: Option<String>,
}

impl MediaCommand {
    pub async fn run(&self, cli: &Cli) -> anyhow::Result<()> {
        match &self.command {
            MediaSubcommand::Subtitle { url, format, language } => {
                self.subtitle(cli, url.as_deref(), format.as_deref(), language.as_deref()).await
            }
            MediaSubcommand::Status { task_id } => {
                self.status(cli, task_id).await
            }
        }
    }

    async fn subtitle(
        &self,
        cli: &Cli,
        url: Option<&str>,
        format: Option<&str>,
        language: Option<&str>,
    ) -> anyhow::Result<()> {
        let ctx = get_context(cli)?;
        let client = create_client(&ctx)?;

        // Load request from file or build from args
        let req = if let Some(input_file) = cli.input.as_deref() {
            let file_req: SubtitleFileRequest = load_request(input_file)?;
            
            let media_url = url.map(|s| s.to_string()).or(file_req.media_url)
                .ok_or_else(|| anyhow::anyhow!("media_url is required"))?;
            
            let format_str = format.or(file_req.format.as_deref()).unwrap_or("srt");
            let subtitle_format = match format_str {
                "vtt" => SubtitleFormat::Vtt,
                "json" => SubtitleFormat::Json,
                _ => SubtitleFormat::Srt,
            };

            let lang_str = language.or(file_req.language.as_deref()).unwrap_or("zh-CN");
            let lang = match lang_str {
                "en-US" => Language::EnUs,
                "ja-JP" => Language::JaJp,
                _ => Language::ZhCn,
            };

            SubtitleRequest {
                media_url,
                format: Some(subtitle_format),
                language: Some(lang),
                enable_translation: false,
                target_language: None,
                callback_url: file_req.callback_url,
            }
        } else {
            let media_url = url.ok_or_else(|| anyhow::anyhow!("media_url is required, use -u flag or -f file"))?;
            
            let format_str = format.unwrap_or("srt");
            let subtitle_format = match format_str {
                "vtt" => SubtitleFormat::Vtt,
                "json" => SubtitleFormat::Json,
                _ => SubtitleFormat::Srt,
            };

            let lang_str = language.unwrap_or("zh-CN");
            let lang = match lang_str {
                "en-US" => Language::EnUs,
                "ja-JP" => Language::JaJp,
                _ => Language::ZhCn,
            };

            SubtitleRequest {
                media_url: media_url.to_string(),
                format: Some(subtitle_format),
                language: Some(lang),
                enable_translation: false,
                target_language: None,
                callback_url: None,
            }
        };

        print_verbose(cli, &format!("Using context: {}", ctx.name));
        print_verbose(cli, &format!("Media URL: {}", req.media_url));

        let result = client.media().extract_subtitle(&req).await?;

        print_success(&format!("Task created: {}", result.task_id));

        output_result(&result, cli.output.as_deref(), cli.json)
    }

    async fn status(&self, cli: &Cli, task_id: &str) -> anyhow::Result<()> {
        let ctx = get_context(cli)?;
        let client = create_client(&ctx)?;

        print_verbose(cli, &format!("Using context: {}", ctx.name));
        print_verbose(cli, &format!("Querying task: {}", task_id));

        let result = client.media().get_subtitle_task(task_id).await?;

        print_success(&format!("Task status: {:?}", result.status));

        output_result(&result, cli.output.as_deref(), cli.json)
    }
}

File: go/cmd/doubaospeech/commands/media.go
package commands

import (
	"context"
	"fmt"
	"time"

	"github.com/spf13/cobra"

	ds "github.com/haivivi/giztoy/go/pkg/doubaospeech"
)

var mediaCmd = &cobra.Command{
	Use:   "media",
	Short: "Media processing service",
	Long: `Media processing service.

Extract subtitles and captions from audio/video files.

Example request file (subtitle.yaml):
  media_url: https://example.com/video.mp4
  language: zh-CN
  format: srt
  enable_translation: true
  target_language: en-US`,
}

var mediaSubtitleCmd = &cobra.Command{
	Use:   "subtitle",
	Short: "Extract subtitles from media",
	Long: `Extract subtitles from audio/video files.

Supports various output formats (srt, vtt, txt).
Can optionally translate subtitles to another language.

Example request file (subtitle.yaml):
  media_url: https://example.com/video.mp4
  language: zh-CN
  format: srt
  enable_translation: true
  target_language: en-US

Examples:
  doubaospeech -c myctx media subtitle -f subtitle.yaml
  doubaospeech -c myctx media subtitle -f subtitle.yaml --json`,
	RunE: func(cmd *cobra.Command, args []string) error {
		if err := requireInputFile(); err != nil {
			return err
		}

		cliCtx, err := getContext()
		if err != nil {
			return err
		}

		client, err := createClient(cliCtx)
		if err != nil {
			return err
		}

		var req ds.SubtitleRequest
		if err := loadRequest(getInputFile(), &req); err != nil {
			return err
		}

		printVerbose("Using context: %s", cliCtx.Name)

		reqCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		task, err := client.Media.ExtractSubtitle(reqCtx, &req)
		if err != nil {
			return fmt.Errorf("subtitle extraction failed: %w", err)
		}

		printSuccess("Subtitle extraction task submitted!")
		printInfo("Task ID: %s", task.ID)
		printInfo("Use 'doubaospeech media status %s' to check status", task.ID)

		result := map[string]any{
			"task_id": task.ID,
			"status":  "submitted",
		}

		return outputResult(result, getOutputFile(), isJSONOutput())
	},
}

var mediaStatusCmd = &cobra.Command{
	Use:   "status <task_id>",
	Short: "Query media task status",
	Long: `Query the status of a media processing task.

Examples:
  doubaospeech -c myctx media status task_12345
  doubaospeech -c myctx media status task_12345 --json`,
	Args: cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		taskID := args[0]

		cliCtx, err := getContext()
		if err != nil {
			return err
		}

		client, err := createClient(cliCtx)
		if err != nil {
			return err
		}

		printVerbose("Using context: %s", cliCtx.Name)
		printVerbose("Querying task: %s", taskID)

		reqCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		status, err := client.Media.GetSubtitleTask(reqCtx, taskID)
		if err != nil {
			return fmt.Errorf("query subtitle task failed: %w", err)
		}

		printInfo("Task ID: %s", status.TaskID)
		printInfo("Status: %s", status.Status)
		if status.Progress > 0 {
			printInfo("Progress: %d%%", status.Progress)
		}
		if status.Result != nil && status.Result.SubtitleURL != "" {
			printSuccess("Subtitle URL: %s", status.Result.SubtitleURL)
		}

		result := map[string]any{
			"task_id":  status.TaskID,
			"status":   status.Status,
			"progress": status.Progress,
		}
		if status.Result != nil {
			result["subtitle_url"] = status.Result.SubtitleURL
			result["duration"] = status.Result.Duration
		}

		return outputResult(result, getOutputFile(), isJSONOutput())
	},
}

func init() {
	mediaCmd.AddCommand(mediaSubtitleCmd)
	mediaCmd.AddCommand(mediaStatusCmd)
}
